### PR TITLE
test(e2e): add registration flow E2E test (POL-7.3)

### PR DESCRIPTION
**Pull request by pcamarajr**
## Summary

This PR implements POL-7.3: E2E test for registration flow. The test validates the complete registration and audio selection flow, verifying that credentials and audio device selections persist across operations.

## Changes

- **New E2E test**: `src-tauri/tests/registration_flow_test.rs`
  - Tests complete flow: initial state verification → registration → credential persistence → audio device listing → device selection → device persistence
  - Uses real SIP client (UDP) for actual SIP server connection
  - Uses mock AudioEngine and CredentialStore for testing
  - Configurable via environment variables (SIP_SERVER_HOST, SIP_SERVER_PORT, SIP_TEST_USER, SIP_TEST_PASSWORD)
  - Can be skipped with SKIP_SIP_INTEGRATION_TESTS env var
- **Documentation updates**:
  - Updated README.md to mark POL-7.3 as complete
  - Updated roadmap to mark POL-7.3 as complete

### How to Test

1. **Setup**:
   ```bash
   git checkout test/pol-7.3-e2e-registration-flow
   cd src-tauri
   ```

2. **Test Steps**:

   - **With SIP server** (requires local Asterisk server):
     ```bash
     # Start SIP server (if not already running)
     npm run docker:start
     
     # Run E2E test
     cargo test --test registration_flow_test -- --ignored
     ```
   
   - **Without SIP server** (test will be skipped):
     ```bash
     SKIP_SIP_INTEGRATION_TESTS=true cargo test --test registration_flow_test
     ```

3. **Expected Behavior**:
   - Test should complete all 6 steps successfully
   - Registration should reach "registered" state
   - Credentials should be saved and loadable
   - Audio devices should be listable and selectable
   - Selected devices should persist

### Test Coverage

- [x] Unit tests: N/A (E2E test)
- [x] Integration tests: E2E test added
- [ ] Manual testing: Requires SIP server setup
- [x] Edge cases: Timeout handling, error states

### Code Quality

- [x] Code follows project conventions
- [x] Functions are focused and single-purpose
- [x] Variable names are descriptive
- [x] No code duplication
- [x] Error handling is appropriate

### Framework Compliance

- [x] Rust code follows idiomatic patterns
- [x] Test follows existing integration test patterns
- [x] Uses proper async/await patterns

### Security

- [x] No hardcoded secrets (uses env vars)
- [x] Test credentials are configurable
- [x] Mock credential store for testing

### Documentation

- [x] README.md updated
- [x] Architecture docs updated
- [x] Test file has comprehensive comments

## Breaking Changes

None

## Related Issues

POL-7.3 (Phase 4.1 E2E test suite)

## Additional Notes

- Test is marked with `#[ignore]` by default - run with `--ignored` flag
- Test requires SIP server for full execution (can be skipped with env var)
- Follows patterns from `sip_transport_integration_test.rs` and service test patterns
- Completes Phase 4.1 implementation

### CODE REVIEW

File: src-tauri/tests/registration_flow_test.rs [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+        !output_devices.is_empty(),
+        "Expected at least one output device"
+    );
+    println!(
+        "DEBUG:[E2E] Found {} output device(s)",
+        output_devices.len()
+    );
+
+    // Step 5: Select Audio Devices
+    println!("DEBUG:[E2E] Step 5: Selecting audio devices");
```
Comment (pcamarajr) [inline src-tauri/tests/registration_flow_test.rs]: **⚠️ Warning:** Potential race condition in credential persistence check

This test waits a fixed 500ms for async credential save to complete (line 480), but credentials are saved via `tokio::spawn` which is fire-and-forget. This could lead to flaky tests on slower systems.

**Recommendation:** Use a polling mechanism with a timeout instead of a fixed sleep:

```rust
// Poll for credentials with timeout
let deadline = tokio::time::Instant::now() + Duration::from_secs(5);
let poll_interval = Duration::from_millis(100);

loop {
    if tokio::time::Instant::now() >= deadline {
        panic!("Timeout waiting for credentials to be saved");
    }
    
    let loaded_creds = credential_store.load(&credential_key).await.unwrap();
    if loaded_creds.is_some() {
        break;
    }
    
    tokio::time::sleep(poll_interval).await;
}
```

This pattern is already used for registration state polling (lines 450-475) and should be applied here for consistency.

**Review by pcamarajr — COMMENTED** [review]
## Code Review Summary

✅ **Overall Assessment:** The PR successfully implements POL-7.3 with a comprehensive E2E test for the registration flow. The test follows existing patterns and properly validates the complete flow from registration to audio device selection.

**Issues Found:**
- 🚫 Blockers: 0
- ⚠️ Warnings: 2
- 💡 Suggestions: 0

**Key Findings:**

1. **Race condition in credential persistence check** (`src-tauri/tests/registration_flow_test.rs:480`)
   - Fixed 500ms sleep may not be sufficient on slower systems
   - Should use polling mechanism (already used elsewhere in the test)

2. **Unclear timeout error messages** (`src-tauri/tests/registration_flow_test.rs:458, 468, 489, 498, 506, 521`)
   - Multiple `.unwrap()` calls on timeout results provide unclear error messages
   - Should handle timeout errors explicitly for better debugging

**Positive Aspects:**

✅ **Code Quality:**
- Test follows existing patterns from `sip_transport_integration_test.rs`
- Well-structured with clear step-by-step flow
- Comprehensive test coverage (6 steps covering registration → credentials → audio)
- Good use of mock implementations (MockAudioEngine, MockCredentialStore)
- Proper debug logging with `DEBUG:[E2E]` prefix

✅ **Architecture Compliance:**
- Uses real SIP client (UDP) for actual server connection
- Properly abstracts audio and credential storage with mocks
- Follows Clean Architecture principles (uses AppState, services, traits)

✅ **Rust & Tauri Compliance:**
- Idiomatic Rust code with proper async/await patterns
- Uses `tokio::time::timeout` appropriately
- Proper error handling with `Result` types
- Test marked with `#[ignore]` for optional execution

✅ **Documentation:**
- README.md correctly updated to mark POL-7.3 as complete
- Roadmap documentation updated correctly
- Test file has comprehensive comments explaining configuration
- PR description is detailed and helpful

✅ **Testing:**
- E2E test covers complete flow: initial state → registration → credential persistence → audio listing → device selection → device persistence
- Configurable via environment variables
- Can be skipped with `SKIP_SIP_INTEGRATION_TESTS` env var
- Follows integration test patterns

**Review Details:**
- Inline comments posted for line-specific issues
- Categories reviewed: Functionality & Correctness, Code Quality & Maintainability, Security, Architecture Compliance, Rust & Tauri Compliance, Testing & Documentation, README & Documentation Updates

**Recommendation:** Address the warnings to improve test reliability and error messages. The PR is otherwise well-implemented and ready to merge after addressing these minor improvements.

File: src-tauri/tests/registration_flow_test.rs [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+    let exists = credential_store.exists(&credential_key).await.unwrap();
+    assert!(exists, "Credentials should exist in credential store");
+    println!("DEBUG:[E2E] Credentials verified in credential store");
+
+    // Step 4: List Audio Devices
+    println!("DEBUG:[E2E] Step 4: Listing audio devices");
+    let audio_service = app_state.audio_service.lock().await;
+    let input_devices = timeout(Duration::from_secs(3), audio_service.list_input_devices())
+        .await
+        .unwrap()
```
[the remark this answers is not in the document]
Comment (pcamarajr) [reply src-tauri/tests/registration_flow_test.rs]: ✅ Resolved in commit 11fb0149bcef3bd6c5c6248c85b4379f8ef85e9a
